contracts/GasStationExact.sol

```diff
@@ -0,0 +1,287 @@
+    bool isActive;
+    uint96 minBalanceWei;
+    uint96 minTopUpAmountWei;
+    uint56 lastTopUpTimestamp; // enough space for 2 trillion years
+  }
+
+  address private s_keeperRegistryAddress;
+  uint256 private s_minWaitPeriodSeconds;
+  address[] private s_watchList;
+  mapping(address => Target) internal s_targets;
+
+  /**
+   * @param keeperRegistryAddress The address of the keeper registry contract
+   * @param minWaitPeriodSeconds The minimum wait period for addresses between funding
+   */
+  constructor(address keeperRegistryAddress, uint256 minWaitPeriodSeconds) ConfirmedOwner(msg.sender) {
+    setKeeperRegistryAddress(keeperRegistryAddress);
+    setMinWaitPeriodSeconds(minWaitPeriodSeconds);
+  }
+
+  /**
+   * @notice Sets the list of addresses to watch and their funding parameters
+   * @param addresses the list of addresses to watch
+   * @param minBalancesWei the minimum balances for each address
+   * @param topUpAmountsWei the amount to top up each address
+   */
+  function setWatchList(
+    address[] calldata addresses,
+    uint96[] calldata minBalancesWei,
+    uint96[] calldata topUpAmountsWei
+  ) external onlyOwner {
+    if (addresses.length != minBalancesWei.length || addresses.length != topUpAmountsWei.length) {
+      revert InvalidWatchList();
+    }
+    address[] memory oldWatchList = s_watchList;
+    for (uint256 idx = 0; idx < oldWatchList.length; idx++) {
+      s_targets[oldWatchList[idx]].isActive = false;
+    }
+    for (uint256 idx = 0; idx < addresses.length; idx++) {
+      if (s_targets[addresses[idx]].isActive) {
+        revert DuplicateAddress(addresses[idx]);
+      }
+      if (addresses[idx] == address(0)) {
+        revert InvalidWatchList();
+      }
+      if (topUpAmountsWei[idx] == 0) {
+        revert InvalidWatchList();
+      }
+      s_targets[addresses[idx]] = Target({
+        isActive: true,
+        minBalanceWei: minBalancesWei[idx],
+        minTopUpAmountWei: topUpAmountsWei[idx],
+        lastTopUpTimestamp: 0
+      });
+    }
+    s_watchList = addresses;
+  }
+
+  /**
+   * @notice Gets a list of addresses that are under funded
+   * @return list of addresses that are underfunded
+   */
+  function getUnderfundedAddresses() public view returns (address[] memory) {
+    address[] memory watchList = s_watchList;
+    address[] memory needsFunding = new address[](watchList.length);
+    uint256 count = 0;
+    uint256 minWaitPeriod = s_minWaitPeriodSeconds;
+    uint256 balance = address(this).balance;
+    Target memory target;
+    for (uint256 idx = 0; idx < watchList.length; idx++) {
+      target = s_targets[watchList[idx]];
+      if (watchList[idx].balance < target.minBalanceWei) {
+        uint256 delta = target.minBalanceWei - watchList[idx].balance;
+        if (
+          target.lastTopUpTimestamp + minWaitPeriod <= block.timestamp &&
+          balance >= delta &&
+          watchList[idx].balance <= target.minBalanceWei - target.minTopUpAmountWei
```

Comment:
I'd do `delta >= target.minTopUpAmountWei` but it's fine

Reply:
better readability i agree; 02ffa2c